### PR TITLE
tests: tests fixes for sru validation

### PR DESCRIPTION
**Pull request by sergiocazzolato**
These fixes are needed to fix sru validation where the snap snap is not used and we use the snapd deb from proposed instead. 

The change include fixes for tests:

- tests/main/apparmor-batch-reload/task.yaml
- tests/main/apparmor-prompting-flag-restart/task.yaml
- tests/main/apparmor-prompting-integration-tests/task.yaml
- tests/main/apparmor-prompting-snapd-startup/task.yaml
- tests/main/interfaces-requests-activates-handlers/task.yaml
- tests/main/interfaces-snap-interfaces-requests-control/task.yaml
- tests/main/remote-home/task.yaml
- tests/main/store-state/task.yaml

### CODE REVIEW

File: tests/main/remote-home/task.yaml [hunk 1]
```diff
@@ -108,9 +108,13 @@ prepare: |
     # Later on, restart snapd and ensure that nfs/cifs workaround is gone.
     # This cleanup handler is registered before we mount the cifs file system.
     if [ "$(snap debug confinement)" = strict ]; then
-        # We are testing on Ubuntu where we know that reexec is active and we
-        # use an internal apparmor userspace stack.
-        tests.cleanup defer test ! -e /var/lib/snapd/apparmor/snap-confine.internal/nfs-support
+        if tests.info is-reexec-in-use; then
```
Comment (bboozzoo) [inline tests/main/remote-home/task.yaml:111]: thanks!

File: tests/main/apparmor-prompting-flag-restart/task.yaml [hunk 2]
```diff
@@ -14,12 +14,18 @@ systems:
   - ubuntu-core-*
 
 prepare: |
+    if not tests.info is-reexec-in-use && os.query is-ubuntu 22.04; then
+        tests.exec skip-test "Ubuntu 22.04 kernel doesn't support prompting" && exit 0
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply tests/main/apparmor-prompting-flag-restart/task.yaml:18]: 16, 18, 20 have kernels which don't support prompting. So they never support it. The problem is that the 22 kernel is new enough to support prompting, but the apparmor parser is not, *unless* we're using snapd-vendored apparmor, which we usually are. The problem is that when we backport the snapd deb and it's >= the latest/stable snapd snap version, it gets prioritized, and runs without reexec, so we end up not supporting prompting.

File: tests/main/apparmor-prompting-snapd-startup/task.yaml [hunk 3]
```diff
@@ -11,12 +11,18 @@ systems:
   - ubuntu-2*
 
 prepare: |
+    if not tests.info is-reexec-in-use && os.query is-ubuntu 22.04; then
+        tests.exec skip-test "Ubuntu 22.04 kernel doesn't support prompting" && exit 0
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply tests/main/apparmor-prompting-snapd-startup/task.yaml:15]: But I'd also prefer to add this to existing check on line 113 rather than skip the test for this edge case.

File: tests/main/interfaces-requests-activates-handlers/task.yaml [hunk 4]
```diff
@@ -8,16 +8,24 @@ systems:
   - ubuntu-2*
 
 prepare: |
+    if not tests.info is-reexec-in-use && os.query is-ubuntu 22.04; then
+        tests.exec skip-test "Ubuntu 22.04 kernel doesn't support prompting" && exit 0
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply tests/main/interfaces-requests-activates-handlers/task.yaml:12]: Same here though, I'd rather this check be part of the existing one on line 36.

**Review by olivercalder — CHANGES_REQUESTED** [review]
The problem with 22.04 without re-exec is that the system parser doesn't support prompting, not the kernel; the kernel does support it.

There are existing checks for expected prompting support in the `execute` sections of the prompting-related tests, and these validate that if we don't expect support, prompting is indeed not supported, so I'd rather integrate this new check into those existing cases, so it's more consistent and we know if/when that support changes (e.g. because of AppArmor parser being backported).

File: tests/main/apparmor-prompting-integration-tests/task.yaml [hunk 5]
```diff
@@ -36,28 +36,39 @@ environment:
...
     echo "outstanding prompts:"
     snap debug api "/v2/interfaces/requests/prompts?user-id=$TEST_UID" || true
     echo "rules:"
     snap debug api "/v2/interfaces/requests/rules?user-id=$TEST_UID" || true
 
 execute: |
+    tests.exec is-skipped && exit 0
+
     echo "Precondition check that snapd is active"
     systemctl is-active snapd.service snapd.socket
```
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/main/apparmor-prompting-integration-tests/task.yaml:74]: so, all the checks to validate that apparmor prompting works in the supported systems and it doesn't in the unsupported will me moved to a new test with just one variant to avoid rechecks.
I'll push a following pr once this is merged.

File: tests/main/apparmor-prompting-integration-tests/task.yaml [hunk 6]
```diff
@@ -74,7 +85,7 @@ execute: |
 
         exit 0
     fi
-
+
```
Comment (olivercalder) [inline tests/main/apparmor-prompting-integration-tests/task.yaml:88]:
```suggestion

```

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml [hunk 7]
```diff
@@ -18,10 +18,16 @@ environment:
     PYTHONIOENCODING: utf-8
 
 prepare: |
+    if not tests.info is-reexec-in-use && os.query is-ubuntu 22.04; then
+        tests.exec skip-test "Ubuntu 22.04 kernel doesn't support prompting" && exit 0
```
[the remark this answers is not in the document]
Comment (olivercalder) [reply tests/main/interfaces-snap-interfaces-requests-control/task.yaml:22]: Only other remaining blocker

**Review by olivercalder — APPROVED** [review]
Looks good, pending the final comment fix `s/kernel/AppArmor parser/`

And good idea to consolidate all the prompting support checks into a dedicated spread test, so all of the prompting behavior checks can just run on the systems for which support is expected and not deal with duplicate checks in each one.

File: tests/main/apparmor-prompting-flag-restart/task.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergiocazzolato) [reply tests/main/apparmor-prompting-flag-restart/task.yaml]: We agreed with Oliver on adding a new test in a following pr including the validations and I'll add all the explanations as well.

**Review by ernestl — APPROVED** [review]
Thanks, approving conditionally to blocker comment about kernel not supporting being corrected.

**Review by maykathm — APPROVED** [review]

File: tests/main/interfaces-snap-interfaces-requests-control/task.yaml (hunk 7 above)
Comment (olivercalder) [inline tests/main/interfaces-snap-interfaces-requests-control/task.yaml:22]:
```suggestion
        tests.exec skip-test "Ubuntu 22.04 AppArmor parser doesn't support prompting" && exit 0
```
